x-pack/plugins/proxy/server/proxy.ts

```diff
@@ -0,0 +1,291 @@
+    // should liveness timing be part of configuration?o
+    // mikhael -- talk about changing http service to allow multiple services
+    this.configSubscription = this.config$.subscribe(config => {
+      this.setConfig(config);
+    });
+
+    const config = await this.config$.pipe(first()).toPromise();
+    this.setConfig(config);
+    this.elasticsearch = core.elasticsearch.dataClient$;
+    this.proxyServer = await core.http.createNewServer({ port: config.port });
+  }
+
+  private setConfig(config: ProxyConfig) {
+    this.updateInterval = config.updateInterval || this.updateInterval;
+    this.timeoutThreshold = config.timeoutThreshold || this.timeoutThreshold;
+    this.port = config.port || this.port;
+  }
+
+  public async start(core: CoreStart) {
+    // start up an http service
+    // update es doc to explain that we exist
+    // start updateLivenessMap
+    this.mainLoop();
+  }
+
+  public async stop() {
+    // stop http service
+    if (this.timer) {
+      clearTimeout(this.timer);
+    }
+
+    const nodes = await this.getNodeList();
+    nodes.delete(this.nodeName);
+    await this.updateNodeList(nodes);
+
+    if (this.configSubscription === undefined) {
+      return;
+    }
+
+    this.configSubscription.unsubscribe();
+    this.configSubscription = undefined;
+  }
+
+  public async assignResource(resource: string, data: RoutingNode): Promise<void> {
+    this.routingTable.set(resource, data);
+    const nodes = await this.getNodeList();
+    const currentTime = new Date().getTime();
+    await this.updateNodeList(this.updateLocalNode(nodes, currentTime));
+  }
+
+  public async unassignResource(resource: string) {
+    this.routingTable.delete(resource);
+    const nodes = await this.getNodeList();
+    const currentTime = new Date().getTime();
+    await this.updateNodeList(this.updateLocalNode(nodes, currentTime));
+  }
+
+  public async registerRouter(router: Router) {
+    this.proxyServer!.registerRouter(router);
+  }
+
+  public async proxyResource(resource: string) {
+    return (req: Request, h: ResponseToolkit) => {
+      this.proxyRequest(req, h, resource);
+    };
+  }
+
+  public async proxyRequest(req: Request, h: ResponseToolkit, resource?: string) {
+    const url = new URL(req.url.toString());
+    resource = resource || url.pathname;
+    const node = this.routingTable.get(resource);
+
+    if (!node) {
+      this.log.debug(`No node was found for resource ${resource}`);
+      const res = h.response('Not found');
+      res.code(404);
+      return res;
+    }
+
+    if (node.state !== RouteState.Started) {
+      this.log.warn(
+        `Unable to foward request for ${resource} to ${node.node} because node was ${node.state}`
+      );
+      const res = h.response('Service Unavailable');
+      res.code(503);
+      return res;
+    }
+
+    url.hostname = node.node;
+    try {
+      const { payload } = await this.wreck.get(url.toString(), { json: true });
```

Comment:
we are able to proxy only `get` requests?

Reply:
This needs to be updated so you can pass through other request typers

Reply:
h2o2 is interesting, but ultimately i'm not sure if we can use it since it is predicated on there being a "stable" mapping for the proxy -- it is possible for the proxy for a specific resource to change the upstream server between calls to it, depending on what the code plugin (or other resources) are doing.

i'm likely going to make sure this looks and feels similar to that though!

Reply:
Personally, I'd prefer not using that plugin. It's not well designed, and IMO too tiny to justify boxing ourselves into an implementation we can't change easily. Also, I think we should avoid using any Hapi plugins in general since we want to be able to remove or replace Hapi easily.